Fixed #36235 -- Fixed RelatedManager queryset losing related instance in `get_or_create()` and `update_or_create()`.

JaeHyuckSa opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-36235

#### Branch description
Fixed an issue where calling `get_or_create()` or `update_or_create()` on a queryset obtained from RelatedManager.all() lost the context of the related instance, causing an IntegrityError.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

tim-schilling commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -694,6 +694,8 @@ def _apply_rel_filters(self, queryset):
                 if val is None or (val == "" and empty_strings_as_null):
                     return queryset.none()
             if self.field.many_to_one:
+                queryset._related_instance = self.instance
+                queryset._related_field_name = self.field.name
```
Why did you choose to go this route rather than using the existing values in `queryset._known_related_objects`?

JaeHyuckSa replied:
As you pointed out, I could have utilized queryset._known_related_objects as already mentioned in the issue, but I missed that. Thanks for the great feedback—I have updated the code accordingly.

ngnpope commented:
Thanks for picking this up. One thing I didn't mention on the ticket was that `.get_or_update()` and `.update_or_create()` were overridden on the `RelatedManager` created under `create_reverse_many_to_one_manager()`. The test I provided was for that case...

But there is also `ManyRelatedManager` created under `create_forward_many_to_many_manager()`. Could you add a test to check whether that case is also fixed?

ngnpope commented on django/db/models/query.py:
```diff
@@ -927,6 +927,10 @@ def get_or_create(self, defaults=None, **kwargs):
         Return a tuple of (object, created), where created is a boolean
         specifying whether an object was created.
         """
+        if self._known_related_objects:
+            field, related_objects = next(iter(self._known_related_objects.items()))
+            rel_obj = next(iter(related_objects.values()))
+            kwargs[field.name] = rel_obj
```
I would suggest extracting this to a method as it is duplicated, e.g. `self._possibly_inject_fk_val(kwargs)`

And then you can add a docstring to that explaining what this is all about.

JaeHyuckSa replied:
Thank you! It looks much cleaner.

ngnpope commented on django/db/models/query.py:
```diff
@@ -968,7 +969,7 @@ def update_or_create(self, defaults=None, create_defaults=None, **kwargs):
         if create_defaults is None:
             create_defaults = update_defaults
-
+        kwargs = self._possibly_inject_fk_val(kwargs)
```
```suggestion
        self._possibly_inject_fk_val(kwargs)
```

JaeHyuckSa commented on django/db/models/query.py:
```diff
@@ -1040,6 +1041,17 @@ def _extract_model_params(self, defaults, **kwargs):
+            field, related_objects = next(iter(self._known_related_objects.items()))
+            rel_obj = next(iter(related_objects.values()))
+            kwargs[field.name] = rel_obj
+        return kwargs
```
I missed it. Thank you

JaeHyuckSa commented:
@ngnpope Thank you for taking the time to review this despite your busy schedule. 
Here are my answers to a few of your questions.

> Thanks for picking this up.
> 
> ~One thing I didn't mention on the ticket was that `.get_or_update()` and `.update_or_create()` were overridden on the `RelatedManager` created under `create_reverse_many_to_one_manager()`. The test I provided was for that case... But there is also `ManyRelatedManager` created under `create_forward_many_to_many_manager()`. Could you add a test to check whether that case is also fixed?~ _(Actually, I'm not sure whether there is anything to do in the M2M case. Perhaps worth checking?)

I've added test code to examine this part, and it still returns a QuerySet. While I was able to check using `_known_related_objects` in `RelatedManager`, I'm unsure which approach to take with `ManyRelatedManager`. Could you advise me on the right direction for this?

> 
> I think, looking at `RelatedManager`, that this may also be needed for `.create()`?

I agree with you on this point. I'm planning to add test code for the create() method—could you please advise me on the best location to place this test?
> 
> Also, are there any issues with doing this as regards the database router stuff in the `RelatedManager` that won't be applied in this fix for calling these methods on the `QuerySet`?

I'll need a bit more time to look into this part, so I'll follow up with you later.

charettes commented on django/db/models/query.py:
```diff
@@ -1040,6 +1041,16 @@ def _extract_model_params(self, defaults, **kwargs):
+        from a related manager context via _known_related_objects.
+        """
+        if self._known_related_objects:
+            field, related_objects = next(iter(self._known_related_objects.items()))
+            rel_obj = next(iter(related_objects.values()))
+            kwargs[field.name] = rel_obj
```
Is there a reason we are picking the first field and the first instance for this field? It will work in the test case but `_known_related_objects` has a type of `{field: {pk: Model}}` which can contain multiple fields and multiple instances when Queryset combination is used (see usage of `_merge_known_related_objects`)

https://github.com/django/django/blob/bc1bfe12b613334bd625aeb36fd44af96d186c10/django/db/models/query.py#L2008-L2013

In other words, I don't think we can assume that `_known_related_objects` will only contain a single field and related object as assigned by `create_reverse_many_to_one_manager`.

ngnpope replied:
So from the queryset we lose reference to `self.field` and `self.instance` on the `RelatedManager`. These are used by that manager to automatically pass in the related instance to `.create()`, `.get_or_create()`,  and `.update_or_create()`.

I'd merely seen that we already have access to these on the queryset via `_known_related_objects` and thought that could be a useful starting point for investigation. Also I'd seen:

https://github.com/django/django/blob/8b241f84e25f679c459393dab2947c8354eb01a9/django/db/models/fields/related_descriptors.py#L730-L732

I guess that we'll just have to duplicate this stuff and explicitly set another attribute on `QuerySet` that we can use? Do you have any other ideas? 🤔

JaeHyuckSa commented:
This PR is being worked on by someone else, so I’m closing it.
